BaragonService/src/main/java/com/hubspot/baragon/service/elb/ApplicationLoadBalancer.java

```diff
@@ -29,23 +62,401 @@ public ApplicationLoadBalancer(Optional<ElbConfiguration> configuration,
+      }
+    }
+  }
+
+  private boolean agentShouldRegisterInTargetGroup(String baragonAgentInstanceId, TargetGroup targetGroup) {
+    DescribeTargetHealthRequest targetHealthRequest = new DescribeTargetHealthRequest()
+        .withTargetGroupArn(targetGroup.getTargetGroupArn());
+    List<TargetHealthDescription> targets = elbClient.describeTargetHealth(targetHealthRequest)
+        .getTargetHealthDescriptions();
+
+    boolean shouldRegister = true;
+    for (TargetHealthDescription targetHealthDescription : targets) {
+      if (targetHealthDescription.getTarget().getId().equals(baragonAgentInstanceId)) {
+        shouldRegister = false;
+      }
+    }
+    return shouldRegister;
+  }
+
+  private Collection<String> getSubnetsFromLoadBalancer(LoadBalancer loadBalancer) {
+    List<AvailabilityZone> availabilityZones = loadBalancer.getAvailabilityZones();
+    Set<String> subnetIds = new HashSet<>();
+
+    for (AvailabilityZone availabilityZone : availabilityZones) {
+      subnetIds.add(availabilityZone.getSubnetId());
+    }
+
+    return subnetIds;
+  }
+
+  private Optional<TargetGroup> guaranteeTargetGroupFor(BaragonGroup baragonGroup) {
+    DescribeTargetGroupsRequest targetGroupsRequest = new DescribeTargetGroupsRequest()
+        .withNames(baragonGroup.getName());
+    List<TargetGroup> targetGroups = elbClient.describeTargetGroups(targetGroupsRequest)
+        .getTargetGroups();
+    if (targetGroups.isEmpty()) {
+      LOG.info("No target group set up for BaragonGroup {}. Skipping.", baragonGroup);
+      return Optional.absent();
+    } else {
+      return Optional.of(targetGroups.get(0));
+    }
+  }
+
+  private Collection<TargetDescription> targetsInTargetGroup(TargetGroup targetGroup) {
+    DescribeTargetHealthRequest targetHealthRequest = new DescribeTargetHealthRequest()
+        .withTargetGroupArn(targetGroup.getTargetGroupArn());
+    List<TargetHealthDescription> targetGroupsResult = elbClient
+        .describeTargetHealth(targetHealthRequest)
+        .getTargetHealthDescriptions();
+
+    Collection<TargetDescription> targetDescriptions = new HashSet<>();
+    for (TargetHealthDescription targetHealthDescription : targetGroupsResult) {
+      targetDescriptions.add(targetHealthDescription.getTarget());
+    }
+
+    return targetDescriptions;
+  }
+
+  private Collection<TargetDescription> listOldTargets(BaragonGroup baragonGroup,
+                                                       Collection<TargetDescription> targetsOnGroup,
+                                                       Collection<BaragonAgentMetadata> knownAgents) {
+    Collection<String> knownAgentIds = agentIds(knownAgents);
+
+    Collection<TargetDescription> removableTargets = new HashSet<>();
+    for (TargetDescription targetDescription : targetsOnGroup) {
+      if (! knownAgentIds.contains(targetDescription.getId()) && canDeregisterAgent(baragonGroup, targetDescription.getId())) {
```

Comment:
I think the data you are looking at is slightly different than you are expecting here. Your `agentIds` method below is returning Baragon's agent id (which in this case is the url on which the agent can be hit by BaragonService). What you probably want in the `knwonAgentIds` variable is the instance IDs which should be equal to the target id. Instance ID should also be available in the BaraongAgentMetadata. Maybe rename that list as well to better show what it contains once you update it.